storage/remote/client.go

```diff
@@ -297,21 +315,54 @@ func (c *Client) Read(ctx context.Context, query *prompb.Query) (*prompb.QueryRe
 	if err != nil {
 		return nil, fmt.Errorf("error sending request: %w", err)
 	}
-	defer func() {
-		io.Copy(io.Discard, httpResp.Body)
-		httpResp.Body.Close()
-	}()
-	c.readQueriesDuration.Observe(time.Since(start).Seconds())
-	c.readQueriesTotal.WithLabelValues(strconv.Itoa(httpResp.StatusCode)).Inc()
 
-	compressed, err = io.ReadAll(httpResp.Body)
-	if err != nil {
-		return nil, fmt.Errorf("error reading response. HTTP status code: %s: %w", httpResp.Status, err)
+	if httpResp.StatusCode/100 != 2 {
+		// Make an attempt at getting an error message.
+		body, _ := io.ReadAll(httpResp.Body)
+		_ = httpResp.Body.Close()
+
+		return nil, fmt.Errorf("remote server %s returned http status %s: %s", c.urlString, httpResp.Status, string(body))
 	}
 
-	if httpResp.StatusCode/100 != 2 {
-		return nil, fmt.Errorf("remote server %s returned HTTP status %s: %s", c.urlString, httpResp.Status, strings.TrimSpace(string(compressed)))
+	contentType := httpResp.Header.Get("Content-Type")
+
+	switch {
+	case strings.HasPrefix(contentType, "application/x-protobuf"):
+		c.readQueriesDuration.WithLabelValues("sampled").Observe(time.Since(start).Seconds())
+		c.readQueriesTotal.WithLabelValues("sampled", strconv.Itoa(httpResp.StatusCode)).Inc()
+		return c.handleSampledResponse(req, httpResp, sortSeries)
+	case strings.HasPrefix(contentType, "application/x-streamed-protobuf; proto=prometheus.ChunkedReadResponse"):
+		c.readQueriesDuration.WithLabelValues("chunked").Observe(time.Since(start).Seconds())
+		// We copy cancel here so we can nil out the original and prevent the context from being cancelled as soon as
+		// Read returns.
+		cancelCopy := cancel
+		cancel = nil
+
+		ss := c.handleChunkedResponse(httpResp, query.StartTimestampMs, query.EndTimestampMs, func(err error) {
+			code := strconv.Itoa(httpResp.StatusCode)
+			if err != io.EOF {
+				code = "aborted_stream"
+			}
+			c.readQueriesTotal.WithLabelValues("chunked", code).Inc()
```

Comment:
Something is off - are we sure this "cancel" func will be called on successful chunk response stream? Quick look on codec tells me - no 🤔

Reply:
I think it will be called on successful chunk response stream + exhaustion of the `chunkedSeriesSet`; when that happens, [this line](https://github.com/prometheus/prometheus/blob/d082eec14f02a603260e23e86b7aab4faf38934a/storage/remote/codec.go#L570) will return an `io.EOF` error and that error will get passed into the cancel function [here](https://github.com/prometheus/prometheus/blob/d082eec14f02a603260e23e86b7aab4faf38934a/storage/remote/codec.go#L578).